README.md
# NanoChat
A simple socket chat written in python  

#### Instructions

1. nanoclient.py acts as a normal client that can connect to the main server which then forwards messages to their recipients.  
  
2. nanoserver.py is a server to which multiple clients can connect to talk to each other  
  
#### Features
- server names and client usernames
- server password
- online users listing
- encryption
- ...


changes.txt
0.4
- Bugfixes and optimizations
- Added AES encryption (requires pycrypto)

0.3.6
- Fixed KeyError when connecting with two clients at the same time on the same ip
- Fixed and covered some exceptions and smaller bugs
- PEP8 coding style fixed

0.3.5
- More exceptions covered
- You are now alerted if you try to send a message to a user that is not connected
- Fixed not being able to connect twice without restarting after disconnecting or inputting a wrong password

0.3.4
- Added password system
- Added server naming system
- Bugs fixed

0.3.3
- Fixed bugs
- Added exceptions
- Stylized responses
- Stuff

0.3.2
- friends.txt has been removed (useless)
- files have been renamed
- pretty much a whole new system (featuring a "master server")

client.py
"""Nano Python Client"""

__version__ = "0.3.1"
__author__ = "DefaltSimon"

import socket,sys

def excepthook(exctype, value, traceback):
    if exctype == ConnectionResetError:
        print("Client disconnected.")
        return
    else:
        sys.__excepthook__(exctype, value, traceback)
sys.excepthook = excepthook

class PythonChat:
    def __init__(self):
        self.sock = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
        self.connected = []
    def connect(self,ip,port):
        try:
            self.sock.connect((ip,port))
            self.connected.append(ip)
        except ConnectionRefusedError:
            print("Client refused to connect.")
        except OSError:
            print("Cannot connect twice!")
    def closesocket(self):
        self.sock.close()
    def sendmsg(self,msg):
        self.sock.sendall(str(msg).encode("utf-8"))
    def getdata(self):
        data = self.sock.recv(4096)
        return data

def start():

    closed = False
    validresp = None
    chat = PythonChat()

    print("Nano Chat " + __version__ + "\n------------------")

    # One socket for one connection ONLY!
    #with open("friends.txt","r") as file:
    #    file = file.readlines()
    #    users = []
    #    for c, line in enumerate(file):
    #        if line.startswith("#"):
    #            continue
    #
    #        thing = line.strip("\n").split("::")
    #        host = thing[0]
    #        port = thing[1]
    #        alias = thing[2]
    #        users.append([host,port,alias])
    #    print(users)
    #
    #for host, port, alias in users:
    #    chat.connect(host,int(port))
    #    print("Connected to " + host + " (" + alias + ")")
    #print("Initialization complete.")

    print("What do you want to do? type help for a list of commands")
    while True:
        option = input(">")

        if option == "help":
            print("Commands: connect, send, close, chat, stop")

        elif option == "connect" or option == "cn":
            if closed is True:
                del chat
                chat = PythonChat()
            if chat.connected:
                print("You are already connected to one client. This version does not support multiple client connections.")
            try:
                ip = input("Host:")
                if not ip:
                    ip = "localhost"
                port = input("Port (default 420):")
                if not port:
                    port = 420
                if ip in chat.connected:
                    print("You are already connected to this ip.")
                    continue
                chat.connect(ip,int(port))
            except ValueError:
                print("Incorrect input")
            resp = bytes(chat.getdata()).decode("utf-8")

            #if resp.startswith("Nano Server") and resp.endswith("OK"):
            #    validresp = "Valid"
            #else:
            #    validresp = "Not valid"
            #
            #print("Response: " + resp + " (" + str(validresp) + ")")
            print("Connected to " + str(ip))

        elif option == "close":
            chat.closesocket()
            closed = True
            chat.connected = []
            print("Closed the connection.")

        elif option == "connections":
            print(chat.connected)

        elif option == "stop":
            print("Disconnecting from servers...")
            chat.closesocket()
            print("Done. Bye!")
            exit()

        elif str(option).startswith("chat"):
            if str(option)[len("chat"):] != "":
                msg = str(option)[len("chat")+1:]
            else:
                msg = input("Message:")
            chat.sendmsg(msg)
            print("msg: " + msg)

start()

chatserver.py
"""Nano Python Server"""

__version__ = "0.3.1"
__author__ = "DefaltSimon"

import socket,sys,threading

def excepthook(exctype, value, traceback):
    if exctype == ConnectionResetError:
        print("Client disconnected.")
        return
    else:
        sys.__excepthook__(exctype, value, traceback)
sys.excepthook = excepthook

hostd, portd = "",420

def gotmsg(user,content):
    if not content:
        content = None
    print("Message from " + user + ": " + str(content))

class PythonChat:
    def __init__(self,host,port,users):
        self.wasconnected = []
        self.users = users

        self.sock = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
        self.sock.setsockopt(socket.SOL_SOCKET,socket.SO_REUSEADDR, 1)
        self.sock.setblocking(1)

        self.sock.bind((host, port))
        self.sock.listen()
    def waitforfirstconn(self):
        while True:
            conn, addr = self.sock.accept()
            self.wasconnected.append(str(addr[0]))
            conn.send("Nano Server 0.1 OK".encode("utf-8"))
            try:
                currentuser = str(self.users.get(addr[0])[1])
            except TypeError:
                currentuser = "no user defined"
            print("Connected from " + str(addr[0]) + " (" + currentuser + ")")

            anotherthread = threading.Thread(target=self.waitformsg,args=[conn,addr])
            anotherthread.daemon = True
            anotherthread.start()
    def waitformsg(self,conn1, addr1):
        msg = None
        try:
            msg = bytes(conn1.recv(4096)).decode("utf-8")
        except ConnectionResetError:
            print(str(addr1[0]) + " disconnected.")
        try:
            gotmsg(self.users.get(addr1[0])[1],msg)
        except TypeError:
            gotmsg("somebody",msg)
    def connect(self,ip,port):
        self.sock.connect((ip,int(port)))
    def closesocket(self):
        self.sock.close()
    def getdata(self):
        conn, addr = self.sock.accept()
        data = conn.recv(4096)
        print(data)


users = None
print("Nano Chat Server " + __version__ + "\n------------------------")
try:
    with open("friends.txt","r") as file:
        file = file.readlines()
        users = {}
        for c, line in enumerate(file):
            if line.startswith("#"):
                continue

            thing = line.strip("\n").split("::")
            host1 = thing[0]
            port1 = thing[1]
            alias = thing[2]
            users[host1] = [port1, alias]
except IndexError:
    print("Error while parsing friends.txt, quiting.")
    exit()

chat = PythonChat(hostd,portd,users)

chat.waitforfirstconn()

nanoserver.py
# coding=utf-8
"""Nano Python Server"""

__version__ = "0.4"
__author__ = "DefaltSimon"

import socket
import sys
import threading
import json
import configparser

from random import randint
from Crypto.Cipher import AES

def excepthook(exctype, value, traceback):  # Custom exception hook

    if exctype == ConnectionResetError:
        print("Client disconnected.")
        return

    else:
        sys.__excepthook__(exctype, value, traceback)  # If no filters are to be applied, apply default exception hook

sys.excepthook = excepthook  # Hook it!


# 'Constants'

PADDING = 16
KEY_LENGTH = 16
padchar = "="

hostd = ""  # Represents localhost
portd = 420  # Outgoing port
defclientportd = 421  # Incoming port

# Utilities
def pad(key):
    # Pads a string so it can be encrypted
    if len(key) == PADDING:
        return key

    else:
        fill = PADDING - len(key) % PADDING
        return key + padchar * fill

def gotmsg(user, content, to):
    if not content:
        return

    print("Message from " + user + " to " + str(to) + ": " + str(content))

def error(content):
    if not content:
        return

    print("[ERROR] {}".format(content))

# Exceptions
class InvalidParameters(Exception):
    def __init__(self, *args, **kwargs):
        pass

class PythonChat:
    def __init__(self,host,port):
        if not isinstance(host, str) or not isinstance(port, int):
            raise InvalidParameters

        self.users = {}
        self.ips = {}

        self.key = None  # 16-bit key
        self.encryption = None  # AES

        self.servername = ""
        self.password = ""

        self.pending = {}

        self.sock = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
        self.sock.setsockopt(socket.SOL_SOCKET,socket.SO_REUSEADDR, 1)
        self.sock.setblocking(1)

        self.sock.bind((host, port))
        self.sock.listen(5)
    def setservername(self,name):
        self.servername = str(name)

    def setpassword(self,password):
        self.password = str(password)

    def generatekey(self, length):
        key = ""
        for rn in range(length):
            # Random char 0 - 9
            key += str(randint(0, 9))

        if len(key) == length:
            self.key = key
            return key

        else:
            raise InvalidParameters


    def encrypt(self,content):

        if not self.encryption:
            self.setupenc()

        return self.encryption.encrypt(pad(str(content)))

    def setupenc(self):
        if not self.key:
            self.generatekey(KEY_LENGTH)

        if not self.encryption:
            self.encryption = AES.new(self.key)

    def verifyuser(self, username, addr):
        try:
            if self.ips[username] == addr:
                return True
            else:
                return False
        except KeyError:
            return False

    def start(self):

        while True:
            conn, addr = self.sock.accept()

            data = bytes(conn.recv(4096)).decode("utf-8")
            #print("Data: " + data)
            initial = json.loads(data)

            if initial["type"] == "initial":
                if self.password:
                    self.pending[str(addr[0])] = initial["username"]
                else:
                    self.users[str(addr[0])] = initial["username"]
                    self.ips[initial["username"]] = str(addr[0])

            else:
                print("Error: Not an initial connection")
                continue

            if self.password:
                raw = {
                    "type": "initialresponse",
                    "code":"OK","servername":self.servername,
                    "password":"required",
                    "key": self.key
                }

            else:
                raw = {"type": "initialresponse",
                       "code":"OK","servername":self.servername,
                       "key":self.key
                       }
            senc = json.dumps(raw).encode("utf-8")

            conn.send(senc)

            try:
                currentuser = str(self.users.get(addr[0]))
            except TypeError:
                currentuser = "error"

            if self.password:
                print("Requesting password from {} ({})".format(addr[0], initial["username"]))
            else:
                print("Connected from {} ({})".format(addr[0], currentuser))

            anotherthread = threading.Thread(target=self.waitformsg,args=[conn,addr])
            anotherthread.daemon = True
            anotherthread.start()
    def waitformsg(self, conn1, addr1):
        while True:

            try:
                demsg = bytes(conn1.recv(4096))

                if not demsg:
                    continue

                # Try to decrypt and/or decode
                demsg = json.loads(str(self.encryption.decrypt(demsg).decode("utf-8")).rstrip("="))

                username1 = str(demsg["username"])

                # Beacuse of multiple threads, this checks if the message received should even be processed
                if self.verifyuser(demsg["username"], addr1[0]) or (self.pending[addr1[0]] == username1):
                    sock2 = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
                    sock2.setsockopt(socket.SOL_SOCKET,socket.SO_REUSEADDR, 1)
                    sock2.setblocking(1)

                    sock3 = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
                    sock3.setsockopt(socket.SOL_SOCKET,socket.SO_REUSEADDR, 1)
                    sock3.setblocking(1)

                    if demsg["type"] == "getclients":
                        sock2.connect((addr1[0],defclientportd))
                        raw = self.encrypt(json.dumps({"type":"clients",
                                          "content":self.users}))
                        sock2.send(raw)

                        print("{} requested clients list.".format(demsg["username"]))

                    elif demsg["type"] == "connectiontest":
                        sock2.connect((addr1[0],defclientportd))
                        raw = self.encrypt(json.dumps({"type":"connectiontest",
                                "content":"ok"}))
                        sock2.send(raw)
                        print("{} tested connection.".format(demsg["username"]))

                    elif demsg["type"] == "disconnect":
                        print("{} ({})  disconnected.".format(addr1[0], self.users[addr1[0]]))

                        try:
                            self.users.pop(str(addr1[0]))
                        except KeyError:
                            pass

                        return  # Exits this thread (representing a connection to a client)

                    elif demsg["type"] == "password":
                        if self.pending[addr1[0]] == username1:
                            # Password check

                            if self.password == demsg["content"]:
                                raw = {"type":"passwordresponse",
                                       "wascorrect":True,
                                       "key":self.key
                                       }

                                self.users[str(addr1[0])] = username1
                                self.ips[username1] = str(addr1[0])

                                quit = False
                                print("Connected from " + str(addr1[0]) + " (" + username1 + ")")

                            else:
                                raw = {"type":"passwordresponse",
                                       "wascorrect":False
                                       }

                                quit = True
                                print(str(addr1[0]) + " (" + username1 + ") failed to connect: password incorrect")

                            senc = self.encrypt(json.dumps(raw))
                            sock2.connect((addr1[0],defclientportd))
                            sock2.send(senc)

                            if quit:
                                return  # Exits this thread (representing a connection to a client)

                    if "sendto" in demsg:

                        try:
                            gotmsg(self.users.get(addr1[0]),demsg["content"],demsg["sendto"])

                        except TypeError:
                            gotmsg("somebody",demsg["content"],demsg["sendto"])

                        ipadd = self.ips.get(str(demsg["sendto"]))

                        if ipadd is None:
                            sock2.connect((addr1[0], defclientportd))
                            raw = self.encrypt(json.dumps({"type":"failedmsg",
                                              "failedto":demsg["sendto"],
                                              "reason":"No user with that name"}))
                            sock2.send(raw)

                        else:
                            sock3.connect((addr1[0], defclientportd))
                            raw = self.encrypt(json.dumps({"type":"successfulmsg"}))
                            sock3.send(raw)

                            sock2.connect((ipadd, defclientportd))
                            raw = self.encrypt(json.dumps({"type":"msg",
                                              "from":str(demsg["username"]),
                                              "content":str(demsg["content"])}))
                            sock2.send(raw)


            except ConnectionResetError:
                try:
                    usern = str(self.users[addr1[0]])
                    print("{} ({}) disconnected.".format(usern, addr1[0]))
                except KeyError:
                    error("Could not find a connected client's username.")
                    print("{} disconnected.".format(addr1[0]))

                try:
                    self.users.pop(str(addr1[0]))
                except KeyError:
                    pass

                return

    def closesocket(self):
        self.sock.close()

length = len("Nano Chat Server "  + __version__)
print("Nano Chat Server" + __version__ + "\n" + str( "-" * length))  # Banner

config = configparser.ConfigParser()
config.read("serversettings.ini")

usepass = config.getboolean("Settings","usepassword")
password = config.get("Settings","password")
usename = config.getboolean("Name","usename")
servernam = config.get("Name","servername")

chat = PythonChat(hostd, portd)

if usepass:
    chat.setpassword(password)
    print("Password: " + str(password))

if usename:
    chat.setservername(servernam)
    print("Server name: " + str(servernam))

else:
    servernam = input("Server name: ")
    chat.setservername(str(servernam))

generatedkey = chat.generatekey(KEY_LENGTH)  # Generates a 16-bit key

print("Encryption key: " + str(generatedkey))

print(str( "-" * len("Server Name: " + str(servernam))))

chat.setupenc()  # Sets up encryption (generates a 16-bit key if not done so and creates AESCipher)

chat.start()  # Starts the infinite loop

nanoclient.py
# coding=utf-8
"""Nano Python Client"""

__version__ = "0.4"
__author__ = "DefaltSimon"

import socket, sys, json, threading, time
from Crypto.Cipher import AES

def pad(key):

    if len(key) == PADDING:
        return key

    else:
        fill = PADDING - len(key) % PADDING
        return key + padchar * fill

# 'Constants'

PADDING = 16
padchar = "="
hostm = ""
portm = 421

# Outgoing port is usually 420 (server default)
# Incoming is 421

isclosing = False


def excepthook(exctype, value, traceback):
    if exctype == ConnectionResetError:
        global isclosing
        if not isclosing:
            print("Connection to server could not be established or was closed.")
        else:
            isclosing = False
    else:
        sys.__excepthook__(exctype, value, traceback)


sys.excepthook = excepthook


class NoEncryptytionKey(Exception):
    def __init__(self):
        return

class NanoChat:
    """
    The main Nano Chat class, featuring the main functions.
    """
    def __init__(self):
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.connected = ""
        self.servername = ""
        self.username = ""

        self.connectiontest = False
        self.connectiontime = 0

        self.passwordcorrect = False

        self.listenerlock = False
        self.msgsent = False
        self.reason = ""

        self.enckey = None
        self.encryption = None
    def setusername(self, username):
        self.username = str(username)

    def setenckey(self,key):
        self.enckey = pad(key)

        return self.enckey

    def encrypt(self,content):

        if not self.encryption:
            self.encryption = AES.new(self.enckey)

        if not self.enckey:
            raise NoEncryptytionKey

        encrypted = self.encryption.encrypt(pad(str(content)))

        return encrypted

    def connect(self, ip, port):
        try:
            self.sock.settimeout(2)

            try:
                self.sock.connect((ip, port))
                araw = ({"type": "initial", "username": self.username})
                self.sock.send(json.dumps(araw).encode("utf-8"))
                self.connected = str(ip)

            except OSError as err:
                print("Server could not be reached: " + str(err))
                self.connected = ""
                return False
            return True

        except ConnectionRefusedError:
            print("Client refused to connect.")
            return False

    def closesocket(self):
        try:
            global isclosing
            isclosing = True

            raw = ({"type": "disconnect", "username": self.username})
            senc = self.encrypt(json.dumps(raw))
            self.sock.send(senc)
            time.sleep(0.1)
            self.sock.close()
        except OSError:
            pass  # Ignore if the socket is already closed

    def Lock(self):
        self.listenerlock = True

    def listener(self):
        print("started")

        # Listening socket, running in a thread
        sock2 = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock2.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        sock2.setblocking(1)

        sock2.bind((hostm, portm))
        sock2.listen(3)

        while True:

            # Blocking, waits for a connection
            conn, addr = sock2.accept()

            self.listenerlock = True
            try:

                # Decodes, loads into a dictionary
                data = bytes(conn.recv(4096))

                # Try to decrypt and/or decode
                #try:
                #    data = json.loads(data.decode("utf-8"))
                #except UnicodeDecodeError or json.JSONDecodeError:
                data = json.loads(str(self.encryption.decrypt(data).decode("utf-8")).rstrip("="))

                msgtype = data["type"]

                # Resolves type
                if msgtype == "successfulmsg":  # prints Sent msg: content before Message from user: content if you send a message to yourself
                    self.msgsent = True

                elif msgtype == "msg":
                    author = data["from"]
                    content = data["content"]
                    print("Message from " + author + ": " + content)

                elif msgtype == "clients":
                    clients = data["content"]
                    compiledclients = ""
                    for item in clients.items():
                        compiledclients += item[0] + " : " + item[1]
                        compiledclients += "\n"
                    print("\nClients connected to main server: \n" + str(compiledclients).strip("\n"))

                elif msgtype == "connectiontest":
                    if data["content"] == "ok":
                        self.connectiontest = True
                    else:
                        self.connectiontest = True

                elif msgtype == "passwordresponse":
                    # Parses encryption key

                    # self.setenckey(data["key"])

                    if data["wascorrect"]:
                        self.passwordcorrect = True
                    else:
                        self.passwordcorrect = False

                elif msgtype == "failedmsg":
                    self.msgsent = False
                    self.reason = data["reason"]

                self.listenerlock = False

            finally:
                self.listenerlock = False

    def sendmsg(self, loc, msg):
        """Normal message"""
        raw = ({"type": "msg", "content": str(msg), "username": self.username, "sendto": str(loc)})
        senc = self.encrypt(json.dumps(raw))
        self.sock.settimeout(3)
        self.sock.send(senc)
        self.listenerlock = True

    def sendpass(self, passw):
        """Response with the password to finalize connecting"""
        raw = ({"type": "password", "content": str(passw), "username": self.username})
        senc = self.encrypt(json.dumps(raw))
        self.sock.send(senc)
        # self.listenerlock = True

    def getclients(self):
        """Requests clients. Response is caught by listener function"""
        raw = ({"type": "getclients", "username": self.username})
        senc = self.encrypt(json.dumps(raw))
        self.sock.send(senc)

    def testconnection(self):
        """Tests for connection/correct port forwarding. Response is caught by listener function"""
        raw = ({"type": "connectiontest", "username": self.username})
        senc = self.encrypt(json.dumps(raw))
        self.sock.send(senc)

        tries = 5
        count = 0
        print("Testing..", end="")
        while tries != 0:

            if self.connectiontest is not False:
                print("connection ok")
                self.connectiontest = False
                return
            else:
                print(".", end="")

            count += 1
            time.sleep(0.2)
        print("no connection or ping above 1s.")

    def getdata(self):
        if not self.connected:
            return

        data = bytes(self.sock.recv(4096))

        # Try to decrypt and/or decode
        try:
            data = json.loads(data.decode("utf-8"))
        except UnicodeDecodeError:
            data = json.loads(str(self.encryption.decrypt(data).decode("utf-8")).rstrip("="))


        return data

def start():
    length = len("Nano Chat " + __version__)
    print("Nano Chat " + __version__ + "\n" + str("-" * length))

    chat = NanoChat()

    print("Starting listener....", end="")
    thr = threading.Thread(target=chat.listener)
    thr.daemon = True
    thr.start()

    username = str(input("Username:"))
    chat.setusername(username)

    print("What do you want to do? type help for a list of commands")
    while True:
        option = input(">")

        if option == "help":
            print("Commands: connect, connection, clients, close, chat, test, stop")

        elif option == "aliases":
            aliases = "Aliases: connect - 'cn', chat - 'c', clients - 'users', connection - 'server', close - 'disconnect','dc', stop - 'exit'"
            print(aliases)

        elif option == "connect" or option == "cn":

            if chat.connected:
                print("You are already connected to one server. Current version does not support multiple client-to-server connections.")
                continue

            ip = None

            try:
                ip = str(input("Host:"))
                if not ip:
                    ip = "localhost"

                port = input("Port (default 420):")
                if not port:
                    port = 420

                else:
                    try:
                        port = int(port)
                    except ValueError:
                        print("Invalid port.")
                        continue

                if port == 421:
                    print("Port 421 not allowed, reverting to 420...")
                    port = 420
                if ip in chat.connected:
                    print("You are already connected to this ip.")
                    continue

                cn = chat.connect(ip, port)
                if not cn:
                    continue

            except ValueError as err:
                print("Incorrect input" + str(err))
                continue

            try:
                resp = chat.getdata()
                chat.servername = resp["servername"]

                chat.setenckey(str(resp["key"]).rstrip("="))
                try:
                    passwn = bool(resp["password"] == "required")
                except KeyError:
                    passwn = False

                if passwn:  # Request password from the user if the server has it
                    password = input("Password: ")
                    chat.sendpass(password)

                    chat.Lock()  # Employ the listener lock
                    isk = False

                    while chat.listenerlock:
                        time.sleep(0.1)
                    if chat.passwordcorrect:
                        print("Correct!")
                        isk = True

                else:
                    isk = True

                if isk:

                    if resp["code"] == "OK":
                        validresp = "response ok"
                    else:
                        validresp = "response incorrect?!"
                    print("Connected to " + str(ip) + " (" + chat.servername + ")..." + validresp)

                else:
                    print("Incorrect password, try again")

                    # Reinitializes the whole NanoChat class to avoid problems.
                    chat.__init__()
                    chat.setusername(username)
                    continue
            except:
                print("No data in response.")

        elif option == "close" or option == "disconnect" or option == "dc":
            if not chat.connected:
                print("You are not connected to a server. Use 'connect' to connect to a server")
                continue

            chat.closesocket()
            print("Disconnected from " + chat.connected)

            chat.__init__()
            chat.setusername(username)

        elif option == "connection" or option == "server":
            if not chat.connected:
                print("You are not connected to a server. Use 'connect' to connect to a server")
                continue

            clients = chat.connected
            print("Server IP: " + clients + "\nName: {}".format(chat.servername))

        elif option == "test" or option == "test connections":
            if not chat.connected:
                print("You are not connected to a server. Use 'connect' to connect to a server")
                continue

            chat.testconnection()

            while chat.listenerlock:
                time.sleep(0.1)

        elif option == "clients" or option == "users":
            if not chat.connected:
                print("You are not connected to a server. Use 'connect' to connect to a server")
                continue
            chat.getclients()

            while chat.listenerlock:
                time.sleep(0.1)

        elif option == "stop" or option == "exit":
            print("Disconnecting from servers...", end="")
            chat.closesocket()
            print("done. Cya!")
            exit()

        elif str(option).startswith("chat") or str(option) == "c":

            if not chat.connected:
                print("You are not connected to a server. Use 'connect' to connect to a server")
                continue

            loca = input("Send to:")
            msg = input("Message:")

            chat.sendmsg(loca, msg)

            while chat.listenerlock:
                time.sleep(0.1)

            if chat.msgsent:
                print("Sent msg: " + msg)
            else:
                print("Failed to send message ({})".format(str(chat.reason)))
                chat.reason = ""


start()
